---
title: "GATC_enrichment_correction"
output: html_document
date: "2023-07-07"
---
```{r setup, include=FALSE}
library(matrixStats)
library(seqinr)
library(pheatmap)
library(rtracklayer)
library(seqinr)
library(GenomicFeatures)
library(dplyr)
setwd("E:/Universidad/MBB/TFM/methylation_motifs/enrichment")
```


PART 1: COUNT OF GATC MOTIFS IN CDS AND INTERGENIC

```{r}
fastaGenome<-read.fasta("NC_000913.3.fasta",seqtype = "DNA",as.string = FALSE) #load fasta sequence from Ecoli
FG<-getSequence(fastaGenome)[[1]] 
annot<-readGFF("GCF_000005845.2_ASM584v2_genomic.gff") #genome annotation GFF format
annot=as.data.frame(annot)

#select coding positions
annot_cds=annot[annot$type=="CDS",]%>%select(seqid, start, end,type,locus_tag)
duplicated_values=annot_cds$locus_tag[duplicated(annot_cds$locus_tag)] #some coding genes have two "alternative transcripts", keep the longes
annot_cds$difference <- annot_cds$end - annot_cds$start
annot_cds <- annot_cds[annot_cds$difference == ave(annot_cds$difference, annot_cds$locus_tag, FUN = max), ]



#select intergenic positions, everything that is not a gene
annot_genes=annot[annot$type=="gene",]%>%select(seqid, start, end)
intergenic <- annot_genes %>% mutate(end_int = lead(start, default = first(start- 1)),start_int = end +1 ) %>% select(seqid, start_int, end_int)%>%add_row(seqid = "NC_000913.3", start_int = 0, end_int = annot_genes$start[1]-1, .before = 1) %>% dplyr::slice(-n())%>%add_row(seqid = "NC_000913.3", start_int = annot_genes$end[dim(annot_genes)[1]], end_int = length(FG))
intergenic=intergenic[-which(intergenic$start_int>=intergenic$end_int),] #eliminating regions that arise from gene that overlap

```

We extract all the sequences annotated as cds or intergen sepatatedly in  and reverse strand

```{r}
sequences_coding <- list() #we create a list with the sequences that correspond to the cds peaks
sequencesNames <- character()
for(i in 1:dim(annot_cds)[1]){
    fromto<-(annot_cds[i,2]:annot_cds[i,3]) 
    tmpSeq <- FG[fromto] #select the sequences in the genome sequence annotation
    sequences_coding[[i]]<-as.SeqFastadna(tmpSeq)
    sequencesNames[i]<-paste("cds_",annot_cds[i,2],"_",annot_cds[i,3],sep="")
}

sequences_intergenic <- list() #we create a list with the sequences that correspond to the cds peaks
sequencesNames_intergenic <- character()
for(i in 1:dim(intergenic)[1]){
    fromto<-(intergenic[i,2]:intergenic[i,3]) 
    tmpSeq <- FG[fromto] #select the sequences in the genome sequence annotation
    sequences_intergenic[[i]]<-as.SeqFastadna(tmpSeq)
    sequencesNames_intergenic[i]<-paste("int_",intergenic[i,2],"_",intergenic[i,3],sep="")
}
```


```{r count 4mers in each type of data cds/ intergenic}
kmer<-4
#forward cds
kmers_cds<-matrix(0,nrow=length(sequences_coding),ncol=4^kmer) #we will have as rows the number of peaks in our positive set (424) and as columns all the possible 4mers done with 4 nucleotides combinations (1024), the matrix will be filled counting how many of each 5mer is present in each peak and in the forward and reverse strad
for(i in 1:length(sequences_coding)){
  kmers_cds[i,]<-as.numeric(seqinr::count(sequences_coding[[i]],wordsize = kmer)) #vector of 1024 4mer count
}
colnames(kmers_cds)<-names(seqinr::count(sequences_coding[[i]],wordsize = kmer)) #assign the colnames as the 5mers
rownames(kmers_cds)<-sequencesNames



#intergenic
kmers_intergenic<-matrix(0,nrow=length(sequences_intergenic),ncol=4^kmer) #we will have as rows the number of peaks in our positive set (424) and as columns all the possible 4mers done with 4 nucleotides combinations (1024), the matrix will be filled counting how many of each 5mer is present in each peak and in the forward and reverse strad
for(i in 1:length(sequences_intergenic)){
  kmers_intergenic[i,]<-as.numeric(seqinr::count(sequences_intergenic[[i]],wordsize = kmer)) #vector of 1024 5mer count
}
colnames(kmers_intergenic)<-names(seqinr::count(sequences_intergenic[[i]],wordsize = kmer)) #assign the colnames as the 5mers
rownames(kmers_intergenic)<-sequencesNames_intergenic

#genome 4mers
kmers_genome<-t(as.matrix(as.numeric(seqinr::count(FG, wordsize = kmer))))
colnames(kmers_genome)<-names(seqinr::count(FG,wordsize = kmer))
rownames(kmers_genome)<-"genome"
```


```{r total GATC}
annot_cds$length=annot_cds$end-annot_cds$start
dimensions_coding=sum(annot_cds$length)
genome_lenght=4641629
percentage_coding=dimensions_coding*100/genome_lenght

intergenic$length=intergenic$end_int-intergenic$start_int
dimensions_intergenic=sum(intergenic$length)
percentage_intergenic=dimensions_intergenic*100/genome_lenght
percentage_noncoding_genes=100-percentage_intergenic-percentage_coding
values_percentages=c(percentage_intergenic,percentage_noncoding_genes,percentage_coding)
labels_percentages=c("Intergenic","non-coding genes","CDS")

percentage_GATC_intergenic=colSums(kmers_intergenic)["gatc"]*100/(colSums(kmers_genome)["gatc"])
percentage_GATC_cds=colSums(kmers_cds)["gatc"]*100/(colSums(kmers_genome)["gatc"])
percentage_non_codinggenes_GATC=100-percentage_GATC_cds-percentage_GATC_intergenic
values_GATC=c(percentage_GATC_intergenic,percentage_non_codinggenes_GATC,percentage_GATC_cds)
labels_GATC=c("Intergenic GATC","non-coding genes GATC","CDS GATC")

library(ggplot2)

# Create a dataframe with values and labels
df <- data.frame(
  category = labels_GATC,
  value = values_GATC
)

# Create a stacked and filled barplot
plot <- ggplot(df, aes(x = "", y = value, fill = category)) +
  geom_bar(stat = "identity", position = "fill") +
  labs(title = "GATC Distribution", x = "", y = "Percentage") +
  theme_minimal() +
  coord_flip()

# Modify legend labels
colors <- c( "#00BA38","#F8766D","#619CFF")
plot <- plot + scale_fill_manual(values = colors,
                                 breaks = df$category,
                                 labels = paste(df$category, ": ", round(df$value,2),"%", sep = ""))

# Display the plot
print(plot)

##Percentage on genome

df <- data.frame(
  category = labels_percentages,
  value = values_percentages
)
plot <- ggplot(df, aes(x = "", y = value, fill = category)) +
  geom_bar(stat = "identity", position = "fill") +
  labs(title = "Genomic proportion", x = "", y = "Percentage") +
  theme_minimal() +
  coord_flip()
colors <- c( "#00BA38","#F8766D","#619CFF")
plot <- plot + scale_fill_manual(values = colors,
                                 breaks = df$category,
                                 labels = paste(df$category, ": ",round(as.numeric(df$value), 2),"%", sep = ""))
print(plot)





#The intergenic part comprises percentage_intergenic=13.30632 and the GATC percentage is 7.299728  percentage_GATC_intergenic
binom.test(as.integer(percentage_GATC_intergenic),100,p=percentage_intergenic/100) #GATC is evenly distributed

log2((percentage_GATC_intergenic/percentage_intergenic)/(percentage_GATC_cds/percentage_coding))
#There is almost 2fold enrichment of GATC sites in CDS.
```



```{r}
peaks_0_0=read.delim("methylation_0-0.bed",header=FALSE,stringsAsFactors = FALSE)
peaks_4_4=read.delim("methylation_4-4.bed",header=FALSE,stringsAsFactors = FALSE)
peaks_6_6=read.delim("methylation_6-6.bed",header=FALSE,stringsAsFactors = FALSE)
peaks_7_7=read.delim("methylation_7-7.bed",header=FALSE,stringsAsFactors = FALSE)
peaks_SRR1536433=read.delim("SRR1536433.bed",header=FALSE,stringsAsFactors = FALSE)

#classify each position on either cds or non-cds
methylated_region=function(P){
  for (i in 1:nrow(P)) {
    P$Category <- ifelse(any(P[i, 4] >= annot_cds$start & P[i, 4] <=
                                     annot_cds$end), "CDS", "non-cds")
  }
  return(table(P$Category))
}


results_0_0=methylated_region(peaks_0_0)
results_4_4=methylated_region(peaks_4_4)
results_6_6=methylated_region(peaks_6_6)
results_7_7=methylated_region(peaks_7_7)
results_SRR1536433=methylated_region(peaks_SRR1536433)  #in all datasets the peaks seem to be in CDS, it is weird I don't know if I formulated correctly




radcoord=2*pi*peaks_0_0$V2/4641629
lm(peaks_0_0$V5~cos(radcoord)+sin(radcoord)) #I don't know how to interpret this results
```